llvm/test/CodeGen/EraVM/overflow.ll

```diff
@@ -0,0 +1,80 @@
+; RUN: llc < %s | FileCheck %s
+
+target datalayout = "E-p:256:256-i8:256:256:256-i256:256:256-S32-a:256:256"
+target triple = "eravm"
+
+declare { i256, i1 } @llvm.uadd.with.overflow.i256(i256, i256)
+declare { i256, i1 } @llvm.usub.with.overflow.i256(i256, i256)
+declare { i256, i1 } @llvm.umul.with.overflow.i256(i256, i256)
+
+; CHECK-LABEL: add_test
+define i256 @add_test(i256 %a, i256 %b, i256 %x, i256 %y) {
+entry:
+  %res1 = call {i256, i1} @llvm.uadd.with.overflow.i256(i256 %x, i256 %y)
+; CHECK:      add!    r3, r4, r{{[0-9]+}}
+; add.ge is add.of reversed
```

Comment:
This is generally a wrong statement. The correct one is that `add!`, `sub!`, `mul!` set flags so if `LT` or `EQ` is set `GT` is reset and vice versa, so for `uaddo`, `umulo`, `usubo` the select condition is indeed reversible.

Reply:
Oh, worse. My statement is wrong too.
The flag is not reversible (and we need to add executable tests here).
https://matter-labs.github.io/eravm-spec/spec.html#AddDefinition says that `lt` is set when overflow, `eq` is set when the result is `0`. So both overflow and `0` result could be produced at the same time, thus `ge` doesn't identify that the overflow hasn't happened.

Reply:
> Could you please navigate me to the code which guarantees that if glue contradicts to data dependency a copy of a node is inserted? We need to make sure that testing is sufficient.

I tried several times and failed. It makes me think it is not easy and straightforward to detect the case you mentioned in our slack discussion. In order to properly handle this "glue contradicts to data dependency" case, I modified the original patch by not calling the `DAG.ReplaceAllUsesOfValueWith` function for the newly created `FoldedArith` instruction, hence we will end up with two arithmetic instructions always, the original one will be kept for generating result, the second one `FoldedArith` will be used for generating flag for SELECT (the GLUE is used here to prevent DAGSchedule from inserting another flag-setting instruction between them).

Case 1: If the result of the first overflow arithmetic isn't used anywhere, then it will be optimized out. The final code is optimal.
Case 2: For the case you mentioned in Slack, we need those two arithmetic instructions anyway, so the final code is expected.
Case 3: This is the sub-optimal case. If there is just one SELECT instruction and no other flag-setting instruction and the result of the overflow arithmetic is used, then we still get two arithmetic instructions with this patch since we are not calling `DAG.ReplaceAllUsesOfValueWith`.

Reply:
Sorry for late reply -- indeed this is correct reasoning.
If we add 1 to $2^{256}-1$ the result will be 0 with an overflow, therefore both LT/OF flag and EQ flag will be set.

Reply:
Patch is updated to avoid inversing overflow LT now.